base/src/main/java/proguard/optimize/kotlin/KotlinLambdaGroupBuilder.java

```diff
@@ -0,0 +1,408 @@
+        }
+        logger.trace("<init> method of lambda class {}: {}",
+                     ClassUtil.externalClassName(lambdaClass.getName()),
+                     ClassUtil.externalFullMethodDescription(lambdaClass.getName(),
+                                                             initMethod.getAccessFlags(),
+                                                             initMethod.getName(lambdaClass),
+                                                             initMethod.getDescriptor(lambdaClass)));
+
+        String oldInitDescriptor = initMethod.getDescriptor(lambdaClass);
+        String newInitDescriptor = KotlinLambdaGroupInitUpdater.getNewInitMethodDescriptor(lambdaClass, initMethod);
+
+        // Check whether an init method with this descriptor exists already
+        ProgramMethod existingInitMethod = (ProgramMethod)lambdaGroup.findMethod(ClassConstants.METHOD_NAME_INIT,
+                                                                                 newInitDescriptor);
+
+        if (existingInitMethod != null)
+        {
+            return existingInitMethod;
+        }
+
+        initMethod.accept(lambdaClass, new MethodCopier(lambdaGroup,
+                                                        ClassConstants.METHOD_NAME_INIT,
+                                                        oldInitDescriptor,
+                                                        AccessConstants.PUBLIC));
+        ProgramMethod newInitMethod = (ProgramMethod)lambdaGroup.findMethod(ClassConstants.METHOD_NAME_INIT,
+                                                                            oldInitDescriptor);
+
+        // Update the <init> descriptor
+        // Add the necessary instructions to entirely new <init> methods
+        newInitMethod.accept(lambdaGroup, this.initUpdater);
+        return newInitMethod;
+    }
+
+    private ProgramMethod copyLambdaInvokeToLambdaGroup(ProgramClass lambdaClass)
+    {
+        logger.trace("Copying invoke method of {} to lambda group {}",
+                     ClassUtil.externalClassName(lambdaClass.getName()),
+                     ClassUtil.externalClassName(this.classBuilder.getProgramClass().getName()));
+
+        // Note: the lambda class is expected to contain two invoke methods:
+        //      - a bridge method that implements invoke()Ljava/lang/Object; for the Function0 interface
+        //      - a specific method that contains the implementation of the lambda
+        // Assumption: the specific invoke method has been inlined into the bridge invoke method, such that
+        // copying the bridge method to th  e lambda group is sufficient to retrieve the full implementation
+        ProgramMethod invokeMethod = getBridgeInvokeMethod(lambdaClass);
+        if (invokeMethod == null)
+        {
+            throw new NullPointerException("No invoke method was found in lambda class " +
```

Comment:
If something goes wrong e.g. an exception is throw here, does it leave the app in a broken state or it just doesn't then merge?